lib/walker.ts

```diff
@@ -670,29 +670,30 @@ class Walker {
   }
 
   stepPatch(record: FileRecord) {
-    const patch = this.patches[record.file];
+    const patches = this.patches[record.file];
 
-    if (!patch) {
+    if (!patches) {
       return;
     }
 
     let body = (record.body || '').toString('utf8');
 
-    for (let i = 0; i < patch.length; i += 2) {
-      if (typeof patch[i] === 'object') {
-        if (patch[i].do === 'erase') {
-          body = patch[i + 1];
-        } else if (patch[i].do === 'prepend') {
-          body = patch[i + 1] + body;
-        } else if (patch[i].do === 'append') {
-          body += patch[i + 1];
+    for (let i = 0; i < patches.length; i += 2) {
+      const patch = patches[i];
+      if (typeof patch === 'object') {
+        if (patch.do === 'erase') {
+          body = patches[i + 1] as string;
+        } else if (patch.do === 'prepend') {
+          body = patches[i + 1] + body;
+        } else if (patch.do === 'append') {
+          body += patches[i + 1];
         }
-      } else if (typeof patch[i] === 'string') {
+      } else if (typeof patch === 'string') {
         // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
         // function escapeRegExp
-        const esc = patch[i].replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+        const esc = patch.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
         const regexp = new RegExp(esc, 'g');
-        body = body.replace(regexp, patch[i + 1]);
+        body = body.replace(regexp, patches[i + 1] as string);
```

Comment:
why do you need to cast as a string here? I think that may be a symptom of an earlier typing problem that should be resolved as part of this refactor.

Reply:
I also can't see how this doesn't hit an array out of bounds exception. The current `for` loop logic is convoluted and should probably be refactored to be readable to future contributors. What is it doing? Can parts be broken out into meaningful function names?

Reply:
I reapproached this refactor from the types instead, it should clarify the intent better now.

Reply:
Didn't think about it on the moment but refactoring this in any way from the types or changing how the loop works will be a breaking change to the `"patches"` API.

Reply:
it seems like you've made it better with your most recent changes though. I do not know what the existing community does with these dictionary entries, if they tend to patch them themselves (in which case, yes, this would break their patches) or if they are just used by pkg itself and can be considered an implementation detail.

I'd argue that these are the latter, and as an implementation detail, can't be relied upon by anyone patching them themselves at runtime to have some sort of stable API that we'd have to semver control.

Do you have any hints about what the docs say about these dictionary entries?

Reply:
I can't find any hint that these dictionary entries should be messed with by users. no "do this to support library XYZ" at all.

Reply:
These are the only "docs" I could find regarding the patches https://github.com/vercel/pkg/wiki/Developers#how-it-works, and that's for the developers of pkg (or more accurately, maintainers). I think you might be right that these are only used internally and by pkg itself, so I'm guessing it's okay to move forward with the recent changes?

Reply:
I think this conversation can be resolved then, thanks for the discussion and thought